src/zeit/retresco/content.py

```diff
@@ -0,0 +1,199 @@
+import UserDict
+import grokcore.component as grok
+import lxml.objectify
+import os.path
+import zeit.cms.interfaces
+import zeit.connector.interfaces
+import zeit.retresco.interfaces
+import zope.component
+import zope.schema.interfaces
+
+
+class Content(object):
+
+    zope.interface.implements(zeit.retresco.interfaces.ITMSContent)
+
+    uniqueId = None
+    __name__ = None
+
+    def __init__(self, data):
+        self._tms_payload = data.get('payload', {})
+        if 'url' in data:
+            self.uniqueId = zeit.cms.interfaces.ID_NAMESPACE + data['url'][1:]
+            self.__name__ = os.path.basename(self.uniqueId)
+        self._build_xml_body()
+        self._build_xml_head()
+
+    def _build_xml_body(self):
+        E = lxml.objectify.E
+        self.xml = E.content()
+        for key, value in self._tms_payload.get('xml', {}).items():
+            self.xml.append(getattr(E, key)(value))
+        for container in ['body', 'teaser']:
+            if container not in self._tms_payload:
+                continue
+            container = getattr(E, container)()
+            self.xml.append(container)
+            for key, value in self._tms_payload[container.tag].items():
+                container.append(getattr(E, key)(value))
+
+    def _build_xml_head(self):
+        """head items must be handled explicitly, because of their structure
+        and type conversion, which we need so they are properly queryable.
+
+        We duplicate knowledge about the XML format here for performance
+        reasons. If we wanted to use vivi APIs we first would have to resolve
+        any referenced ICMSContent; instead we can create XML nodes directly.
+        """
+        if 'head' not in self._tms_payload:
+            return
+        data = self._tms_payload['head']
+
+        E = lxml.objectify.E
+        head = E.head()
+        self.xml.append(head)
```

Comment:
so calling this method appends data to self.xml as a sideeffect... so multiple calls would be cumulative? is this desired? and if so why? and if not, why implement it that way?

Reply:
Öh... die Methode `_build_xml_head` existiert rein um die `__init__` in einzelne, abgeschlossene Teile aufzutrennen. Die ist nicht als general purpose "mich kann aufrufen wer will" gedacht, deshalb ist sie auch _privat benannt.